### ISSUE TITLE
It does not reflect real world data

### ISSUE THREAD

**Issue by nopara73** [6102bitcoin/CoinJoinFees#2]
https://github.com/zkSNACKs/WalletWasabi/issues/2077

According to your theory the above 3.5BTC to reach 300 anonset should had paid 0.9% fees, yet it paid 0.35% both times when it was tested.

**Comment by 6102bitcoin** [comment]
I understand that the payment of network fees is complicated depending on utxo size and change output size, but I was under the impression that the coordinator fee is independent of those variables. Is this inaccurate?

**Comment by AvivMilner** [comment]
No! That is why what @nopara73 said is honest. I can even share screenshots with you of my own re-mixing that cost 262 satoshis!!

**Comment by 6102bitcoin** [comment]
Please can you point me to the spec / code that determines how the fee each UTXO pays is calculated.

**Comment by AvivMilner** [comment]
This makes it annoying for people to discuss the Wasabi fees, because like I mentioned about, they are not deterministic! You cannot predict them until after the CoinJoin. They also change depending on a variety of factors (e.g. what is the current round denomination? how many participants? how much change do you have? What are network fee?

**Comment by AvivMilner** [comment]
https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs

**Comment by AvivMilner** [comment]
Now let's have a moment of honesty and reflection about Samourai fees vs. Wasabi fees.

- Samourai only does 5 person CJ, and users pay a flat fee to enter any amount
- Wasabi has variable participants and variable rates.
- Both Samourai + Wasabi charge users for txn fees, which are variable
- Samourai has free re-mixing (but it takes a very very long time to remix)
- Wasabi has unclear (but lower, sometimes) re-mixing fees, but you can remix at any time.
- Samourai has consistent predictable round denominations
- Wasabi has unpredictable round denominations 

In summary - Samourai is simpler to understand and explain in terms of fees.

**Comment by 6102bitcoin** [comment]
I am interested in comparing fees - I am happy to hear any thoughts on how best to do this. One option which was mentioned on slack was to mix-and-measure which is an option, but not something that I can do regularly thus I am interested in better understanding exactly how the fees are determined so that I can evaluate whether it is possible to improve this script without resorting to using mix-and-measure estimates which will be out of date from the day they are posted. 

To be clear, I am not being intentionally misleading regarding the wasabi fees. This method of calculating fees is what is often used even by @nopara73 in the latest blockdigest "0.0015 (BTC Fee), This would be the first round (Fee) and every round is similar"

**Comment by AvivMilner** [comment]
@6102bitcoin I can tell that you are trying to be a good scientist here and give an accurate prediction for future CoinJoins. I am going to tell you right now, that I can tweak the circumstances and user will pay much more or much less than any prediction you give. So it is really not worth it.

I like to say that users should expect to be roughly ~0.3% as long as they set their wallet to remix until >100 anonset. This is pretty accurate most of the time, and this is the best you will do.

@nopara73 also gave a clear method, but it just takes the noisy variable rate and gives you the average across 50 CoinJoins or so.

**Comment by 6102bitcoin** [comment]
By his method are you referring to mix-and-measure?

Please can you elaborate on how you could tweak the circumstances to make a user pay much more / much less? Is my method the upper bound on the fee?

**Comment by AvivMilner** [comment]
(1) Yes, mix a large amount and measure

(2) Sure, suppose you have 0.108 BTC, but the round denomination is 0.1067 BTC. Because your change is less than 1mbtc, it is not returned to you and instead goes towards fees. In this case you paid triple by accident.

Next, suppose you remix, but the network fees are really high! You will pay more fees at that time.

**Comment by 6102bitcoin** [comment]
(1) I am not sure how the larger denomination mixing comes into this - It could be that this is what is throwing the estimate off. (E.g. A 0.8 BTC UTXO could be mixed in the first round, but this wouldn't have a 300 Anonset). Is it possible to get a breakdown of mixed UTXO amounts and their anonset after mixing the 3.5 BTC @nopara73 ?

(2) Because your change is less than **1mbtc**, it is not returned to you and instead goes towards fees. 

I don't see this rule in the [Coordinator Code](https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs) am I looking in the wrong place?

Next, suppose you remix, but the network fees are really high! You will pay more fees at that time.

**Comment by 6102bitcoin** [comment]
> but the network fees are really high! You will pay more fees at that time.

This fee calculator does not account for network fees - It is on the [to-do list](https://github.com/6102bitcoin/CoinJoinFees/blob/master/README.md) but It has not been done.

**Comment by 6102bitcoin** [comment]
Thinking it through, if you mix a large amount you will have multiple inputs which will be mixing through many rounds. Given what you say about low fees if you mix in consecutive rounds this advantages you significantly the more consecutive rounds you mix for, compared to intermittent round mixing. 

As such, mixing with a high anon-set target over consecutive rounds will give you a lower % fee *per anonset* than you would get; 
- Mixing in non-consecutive rounds
- OR Mixing for fewer rounds

As such, this size amount (3.5 BTC) , being mixed for many consecutive rounds (300 Anonset) will give you something close to the lower bound on what fees reasonably be expected to be.

If, in contrast, someone gets unlucky and mixes a UTXO which is larger than the round denomination but not by enough to get the output paid back (e.g. your 0.108 BTC example) they could pay significantly higher fees.

My estimate is therefore underestimating fees for low anonset mixes which are slightly too large, and overestimating fees for high anonset mixes which mix in consecutive rounds.

Short of mix-and-measure with many different sized amounts & anon-set targets it's hard to know whether it is even possible to give an accurate estimate other than saying that the % fee could be significantly higher for low anonsets and somewhat lower for high anonset amounts.

**Comment by 6102bitcoin** [comment]
To put some numbers to this: 
- The lower bound is actually negative (You get back more than you input), I have no idea how much we are talking here but presumably < 0.1%?
- The upper bound can be ~ 0.9% (From @AvivMilner's example) 

If the rules for the mixing fees were clearly listed I could try running a simulation to see what fees you would expect.

**Comment by AvivMilner** [comment]
(1) The larger amounts are not throwing off the numbers, it's just that there are more coinjoins and so it is a better average.

(2) I am looking in the code, but I cannot find it. It is somewhere near line 300 if I recall

**Comment by yahiheb** [comment]
> (2) Because your change is less than 1mbtc, it is not returned to you and instead goes towards fees.
> 
> I don't see this rule in the Coordinator Code am I looking in the wrong place?

https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/CoinJoin/Coordinator/Rounds/CoordinatorRound.cs#L368

https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/Legal/Assets/LegalDocuments.txt#L77

**Comment by 6102bitcoin** [comment]
@AvivMilner - Is is not likely that mixing in consecutive rounds results in lower average fees being paid than the fees paid in the first round?

**Comment by AvivMilner** [comment]
Yes, @6102bitcoin , if you mix in consecutive rounds, you are almost guaranteed to pay less than the first time you mix.

**Comment by 6102bitcoin** [comment]
So you see - mixing for many many rounds biases the fees down (as opposed to providing a more 'accurate' fee amount).

@nopara73 do you have data for the fees paid per round and the anon-set gained per round?

Then we could plot fee per anonset and you'd see that it drops over time with more remixing.
